### PR TITLE
Controller Input Action Mapping Popout Menus

### PR DESCRIPTION
**Pull request by StephenHodgson**
Overview
---
image
image
image

Notes:
---
- Had to completely replace the default controller profile to make sure all the values were initialized and synchronized correctly across all device types.
- Had to update interaction mappings to make sure devices of the same type had the same exact interaction layout sans the axis and button mappings.
    - As a consequence I had to change the WMR Controller a bit to support it.

Changes
---
- Fixes: #2617
- Fixed the axis constraints for Oculus Remote buttons.
- Fixed an issue with constructors not being updated after the X Axis Inversion addition.

### CODE REVIEW

**Comment by StephenHodgson** [comment]
Already merged it in.

**Comment by SimonDarksideJ** [comment]
OK @StephenHodgson two main contentious queries.

- If we are going to have ALL controllers available by default (tried creating a new profile where I expected no controllers). Then we need the images in the inspector to be a LOT smaller.  I'd preferably like a small grid of Icons (as that is all they are now) rather than the huge images in the inspector.   The call outs are perfectly fine, it's just the inspector one's which are huge for no good reason.

- The Custom editor popout it a great improvement, however, can we have it open with one interaction mapping already available.  UX guidelines simply state (if there is only one thing to do, do it automatically instead of forcing a user to click a button).  When I open up a new custom controller editor window, I have no interactions and a single button to press, better there is one new interaction for the new mapping and the button adds more.

**Comment by StephenHodgson** [comment]
> it's just the inspector one's which are huge for no good reason.

I couldn't get them to look good smaller.
I also couldn't figure out how to renderer them in a grid while also skipping controllers of the same physical type.

> When I open up a new custom controller editor window, I have no interactions and a single button to press, better there is one new interaction for the new mapping and the button adds more.

I wasn't sure what the default generic unity mappings should be. That can be easily addressed in a separate PR

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+                                currentAxisSetting += 1;
+                            }
+
+                            if (invertYAxis.boolValue)
+                            {
+                                currentAxisSetting += 2;
+                            }
+
+                            EditorGUI.BeginChangeCheck();
+                            currentAxisSetting = EditorGUILayout.IntPopup(InvertContent, currentAxisSetting, InvertAxisContent, InvertAxisValues, GUILayout.Width(InputActionLabelWidth));
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: Doesn't look like #2620 was integrated/rewritten correctly. Also not sure why the change in the list was made, causing the display to go from "Invert Y" to "Invert Invert Y".

Before: 
image

After:
image

File: Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+You may use these images solely for referring to the corresponding product in your video game or VR experience (including manuals for users). Otherwise, you may not use these images, or any Oculus trademarks, logos or other intellectual property, including but not limited to use on merchandise or other product such as clothing, hats, or mugs. Do not use the Oculus images in a way that implies a partnership, sponsorship or endorsement; or features Oculus on materials associated with pornography, illegal activities, or other materials that violate Oculus Terms.
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt]: Would the MRTK be correctly categorized under "video game or VR experience"? I'm not sure.

Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt]: Also on that note, where did the rest of the images come from and what are their licenses?

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs [hunk 3]
```diff
@@ -236,500 +139,242 @@ public override void OnInspectorGUI()
...
+                        useCustomInteractionMappings.boolValue = controllerType == typeof(GenericUnityController) ||
+                                                                 controllerType == typeof(GenericOpenVRController);
+                        interactionsList.ClearArray();
+                        serializedObject.ApplyModifiedProperties();
+                        thisProfile.MixedRealityControllerMappingProfiles[i].SetDefaultInteractionMapping(true);
+                        serializedObject.ApplyModifiedProperties();
                     }
 
-                    if ((AxisType)axisType.intValue == AxisType.DualAxis)
+                    if (renderControllerModels && controllerHandedness.intValue != 0)
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: Why can controllers without a handedness not have a model?

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs [hunk 4]
```diff
@@ -236,500 +139,242 @@ public override void OnInspectorGUI()
...
-                            GUILayout.EndVertical();
-                            return;
-                        }
-
-                        EditorGUIUtility.labelWidth = 128f;
-                        EditorGUILayout.PropertyField(useDefaultModel);
-
-                        if (!useDefaultModel.boolValue)
-                        {
-                            EditorGUILayout.PropertyField(controllerModel);
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: Is the controller-specific model override field intentionally no longer displayed for all non-generic controllers?

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs [hunk 5]
```diff
@@ -2,13 +2,13 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.﻿
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: I'm seeing a couple bugs in this inspector, but didn't find the exact line for them.

Repro steps:
1. Scroll down to the generic Open VR Controllers. Do this on the **Right** Hand Controller, as there's a different bug preventing repro on the Left Hand Controller.
2. Try to change the Generic Type to Unity Controller while there's a Handedness selected. It won't change.
3. Switch the handedness to None, and switch to Unity Controller, successfully.
4. Change handedness to anything other than None, successfully.

The bug is either in step 2 or step 4 (or potentially both), depending on whether Unity Controllers support handedness, I suppose.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs [hunk 6]
```diff
@@ -2,13 +2,13 @@
...
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: Another bug I didn't find a line for.

Repro steps:

1. Scroll down to the generic Open VR Controllers. Do this on the **Left** Hand Controller.
2. Try to change the Generic Type to Unity Controller while there's a Handedness selected. It won't change.
3. Switch the handedness to None, and switch to Unity Controller.
4. Observe that the Right Hand Controller has been changed to Unity Controller instead of the Left Hand Controller.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs (hunk 3 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: Because they're not spatial controllers, which won't be rendered in the scene.

File: Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt (hunk 2 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt]: I made all the images, but just in case there was ambiguity I added the license.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt]: I would say so, yes.

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: I was integrated correctly, and actually fixed a few issues. 

I'll remove the duplicate invert.

File: Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset [hunk 7]
```diff
@@ -540,150 +592,150 @@ MonoBehaviour:
       invertXAxis: 0
       invertYAxis: 0
     - id: 1
-      description: Trigger Position
-      axisType: 3
-      inputType: 10
+      description: Spatial Grip
```
Comment (keveleigh) [inline Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset]: Have you checked that HoloLens hand position comes through spatial pointer and not spatial grip? If I'm remembering correctly, the Unity API surfaces that info across the grip type, not the pointer type.

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: Where did the "Invert X Axis?" fields go?

File: Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset (hunk 7 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-SDK/Profiles/DefaultMixedRealityControllerMappingProfile.asset]: I assigned the grip pose action to the Spatial Grip on the HL

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs (hunk 5 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: Great catch, fixed.

File: Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs (hunk 6 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/Profiles/MixedRealityControllerMappingProfileInspector.cs]: Great catch, fixed.

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: Oh I see. On tick, I'll be sure to update it.

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: Great catch. Fixed

**Comment by Jarodshow** [comment]
Names such as "Button.Three" and "Axis1D.PrimaryHandTrigger" aren't super helpful to someone who's maybe not so familiar with XR. Maybe a new field in `MixedRealityInteractionMapping` called `FriendlyName` or `AlternateName` could help this.

Just a thought.

**Comment by StephenHodgson** [comment]
I thought about updating the names, but left them as they're used in the Oculus device manual's as well.

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs (hunk 1 above)
Comment (keveleigh) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: >and actually fixed a few issues.

For future reference, what specific issues did you fix, and are they things we need to keep an eye out for?

File: Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt (hunk 2 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Resources/Textures/Oculus lineart attribution.txt]: Depends on how technical you wanna get.
I never "drew" anything. 😋

I did process the images and make sure they were in a format that was suitable for use, which involved resizing, and editing the images from their original form. 

I added a 5 px stroke and cleaned up the images where there was errors.

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: Take a look at the description of the issue where I laid them out plainly.

Comment (keveleigh) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: Which issue? The only issue mentioned in this PR is the one about adding the pop-up windows.

Are you referring to the constructor updates mentioned in this PR?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: > Fixed an issue with constructors not being updated after the X Axis Inversion addition.

Yes

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs [hunk 8]
```diff
@@ -224,10 +225,588 @@ public static class ControllerMappingLibrary
...
+
+        private static Texture2D oculusRemoteControllerWhite;
+
+        public static Texture2D OculusRemoteControllerWhite
+        {
+            get
+            {
+                if (oculusRemoteControllerWhite == null)
+                {
+                    oculusRemoteControllerWhite = (Texture2D)AssetDatabase.LoadAssetAtPath("Assets/MixedRealityToolkit/_Core/Resources/Textures/OculusRemoteController_white.png", typeof(Texture2D));
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs]: I would comment / remove this code (entire oculus block) while awaiting guidance on images

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs [hunk 9]
```diff
@@ -224,10 +225,588 @@ public static class ControllerMappingLibrary
...
-        #endregion Controller axis mapping configuration
+        private static Texture2D touchControllerLeftWhite;
+
+        public static Texture2D TouchControllerLeftWhite
+        {
+            get
+            {
+                if (touchControllerLeftWhite == null)
+                {
+                    touchControllerLeftWhite = (Texture2D)AssetDatabase.LoadAssetAtPath("Assets/MixedRealityToolkit/_Core/Resources/Textures/OculusControllersTouch_left_white.png", typeof(Texture2D));
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs]: I would comment / remove this code (entire touch controller block) while awaiting guidance on images

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs [hunk 10]
```diff
@@ -224,10 +225,588 @@ public static class ControllerMappingLibrary
...
+
+        private static Texture2D viveWandControllerLeftWhite;
+
+        public static Texture2D ViveWandControllerLeftWhite
+        {
+            get
+            {
+                if (viveWandControllerLeftWhite == null)
+                {
+                    viveWandControllerLeftWhite = (Texture2D)AssetDatabase.LoadAssetAtPath("Assets/MixedRealityToolkit/_Core/Resources/Textures/ViveWandController_left_white.png", typeof(Texture2D));
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs]: I would comment / remove this code (entire vive block) while awaiting guidance on images

**Comment by SimonDarksideJ** [comment]
If we've removed the specific device images, can we have a default image that is shown in its place. The text alone looks horrible.
Something like this icon from the HTK
image

Also, I'm getting some unused references errors in the latest commit.

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs [hunk 11]
```diff
@@ -0,0 +1,755 @@
...
+
+        private Handedness currentHandedness;
+        private SupportedControllerType currentControllerType;
+
+        private Vector2 mouseDragOffset;
+        private GUIStyle flippedLabelStyle;
+        private Texture2D currentControllerTexture;
+        private ControllerInputActionOption currentControllerOption;
+
+        private bool IsCustomController => currentControllerType == SupportedControllerType.GenericOpenVR ||
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: this likely needs to change while we await guidance on controller images

**Comment by StephenHodgson** [comment]
Will do

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs (hunk 8 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs]: No, we should use default images in their place @davidkline-ms as a default until we have agreed new images
Text alone looks darn ugly

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs (hunk 9 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs]: As above

File: Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs (hunk 11 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Inspectors/ControllerPopupWindow.cs]: As above

**Comment by SimonDarksideJ** [comment]
Can you elaborate on your comment @heatherbarry ?

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs (hunk 8 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/ControllerMappingLibrary.cs]: are these JUST for the mini display? if not, where do the control pointers go to on this image?

**Comment by StephenHodgson** [comment]
They can stay blank for now I think.

**Comment by StephenHodgson** [comment]
I might have them fallback to being in an ordered list if that makes more sense. What do you guys think?

**Comment by davidkline-ms** [comment]
Fallback makes the most sense to me

**Comment by StephenHodgson** [comment]
> Fallback makes the most sense to me

To an ordered list, like I suggested?

**Comment by StephenHodgson** [comment]
Done

**Comment by davidkline-ms** [comment]
@StephenHodgson, you sent met the source link for the Motion Controller models (docs.microsoft.com)... what was the source for the xbox controller model?

**Comment by StephenHodgson** [comment]
Do you have official like art you'd rather me use?

**Comment by davidkline-ms** [comment]
> Do you have official like art you'd rather me use?

I would have to check with the xbox team.. I was wondering the source of the image so we can know if it is something we can redistribute.

**Comment by davidkline-ms** [comment]
> I would have to check with the xbox team.

I have an inquiry out to them right now..

**Comment by StephenHodgson** [comment]
I just want to remove any doubt.

**Comment by StephenHodgson** [comment]
Removed the Xbox Images until official ones are provided.
